File: app/src/main/java/per/goweii/wanandroid/module/main/activity/CrashActivity.java
package per.goweii.wanandroid.module.main.activity;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import cat.ereza.customactivityoncrash.CustomActivityOnCrash;
import cat.ereza.customactivityoncrash.config.CaocConfig;
import per.goweii.basic.utils.CopyUtils;
import per.goweii.statusbarcompat.StatusBarCompat;
import per.goweii.wanandroid.R;
import per.goweii.wanandroid.common.WanApp;

/**
 * @author CuiZhen
 * @date 2019/10/20
 * GitHub: https://github.com/goweii
 */
public class CrashActivity extends AppCompatActivity {

    @BindView(R.id.tv_log)
    TextView tv_log;
    @BindView(R.id.tv_show_log)
    TextView tv_show_log;
    @BindView(R.id.iv_bug)
    ImageView iv_bug;
    @BindView(R.id.sv)
    ScrollView sv;

    private Unbinder mUnbinder = null;
    private CaocConfig mCaocConfig;

    private boolean isLogShown = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mCaocConfig = CustomActivityOnCrash.getConfigFromIntent(getIntent());
        if (mCaocConfig == null) {
            finish();
            return;
        }
        WanApp.initDarkMode();
        StatusBarCompat.setIconMode(this, !WanApp.isDarkMode());
        setContentView(R.layout.activity_crash);
        mUnbinder = ButterKnife.bind(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mUnbinder != null) {
            mUnbinder.unbind();
        }
    }

    @OnClick({R.id.tv_restart, R.id.tv_exit, R.id.tv_show_log})
    public void onClick(View v) {
        switch (v.getId()) {
            default:
                break;
            case R.id.tv_show_log:
                String log = tv_log.getText().toString();
                if (TextUtils.isEmpty(log)) {
                    isLogShown = true;
                    tv_show_log.setText("复制日志");
                    tv_show_log.setTextColor(getResources().getColor(R.color.text_main));
                    iv_bug.setVisibility(View.GONE);
                    sv.setVisibility(View.VISIBLE);
                    tv_log.setText(CustomActivityOnCrash.getAllErrorDetailsFromIntent(this, getIntent()));
                } else {
                    CopyUtils.copyText(log);
                    tv_show_log.setText("日志已复制");
                    tv_show_log.setTextColor(getResources().getColor(R.color.text_third));
                    tv_show_log.setEnabled(false);
                }
                break;
            case R.id.tv_exit:
                CustomActivityOnCrash.closeApplication(this, mCaocConfig);
                break;
            case R.id.tv_restart:
                CustomActivityOnCrash.restartApplication(this, mCaocConfig);
                break;
        }
    }
}

File: app/src/main/java/per/goweii/wanandroid/module/main/activity/CrashActivity.kt
package per.goweii.wanandroid.module.main.activity

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.os.Process
import android.view.View
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_crash.*
import per.goweii.ponyo.crash.Crash
import per.goweii.statusbarcompat.StatusBarCompat
import per.goweii.wanandroid.R
import per.goweii.wanandroid.common.WanApp
import java.io.PrintWriter
import java.io.StringWriter
import kotlin.system.exitProcess

class CrashActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        window.decorView.systemUiVisibility =
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.statusBarColor = Color.TRANSPARENT
        WanApp.initDarkMode()
        StatusBarCompat.setIconMode(this, !WanApp.isDarkMode())
        setContentView(R.layout.activity_crash)
        tv_copy_log.setOnClickListener {
            val cm = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            cm.setPrimaryClip(ClipData.newPlainText("error", tv_error.text))
            tv_copy_log.text = "已复制"
        }
        btn_exit.setOnClickListener {
            finish()
            Process.killProcess(Process.myPid())
            exitProcess(10)
        }
        btn_restart.setOnClickListener {
            Crash.restartApp(applicationContext)
            finish()
            Process.killProcess(Process.myPid())
        }
        showLog()
    }

    private fun showLog() {
        val e = intent.getSerializableExtra("error") as Throwable
        val stringWriter = StringWriter()
        val printWriter = PrintWriter(stringWriter)
        e.printStackTrace(printWriter)
        tv_error.text = stringWriter.toString().toDBC()
    }

    private fun String.toDBC(): String {
        val c = this.toCharArray()
        for (i in c.indices) {
            if (c[i].toInt() == 12288) {
                c[i] = 32.toChar()
                continue
            }
            if (c[i].toInt() in 65281..65374) c[i] = (c[i] - 65248)
        }
        return String(c)
    }
}